drivers/sensor/iis2iclx/iis2iclx.c

```diff
@@ -0,0 +1,657 @@
+
+static int iis2iclx_accel_range_to_fs_val(int32_t range)
+{
+	size_t i;
+
+	for (i = 0; i < ARRAY_SIZE(iis2iclx_accel_fs_map); i++) {
+		if (range == iis2iclx_accel_fs_map[i]) {
+			return i;
+		}
+	}
+
+	return -EINVAL;
+}
+#endif
+
+static inline int iis2iclx_reboot(const struct device *dev)
+{
+	struct iis2iclx_data *data = dev->data;
+
+	if (iis2iclx_boot_set(data->ctx, 1) < 0) {
+		return -EIO;
+	}
+
+	/* Wait sensor turn-on time as per datasheet */
+	k_busy_wait(35 * USEC_PER_MSEC);
+
+	return 0;
+}
+
+static int iis2iclx_accel_set_fs_raw(const struct device *dev, uint8_t fs)
+{
+	struct iis2iclx_data *data = dev->data;
+
+	if (iis2iclx_xl_full_scale_set(data->ctx, fs) < 0) {
+		return -EIO;
+	}
+
+	data->accel_fs = fs;
+
+	return 0;
+}
+
+static int iis2iclx_accel_set_odr_raw(const struct device *dev, uint8_t odr)
+{
+	struct iis2iclx_data *data = dev->data;
+
+	if (iis2iclx_xl_data_rate_set(data->ctx, odr) < 0) {
+		return -EIO;
+	}
+
+	data->accel_freq = iis2iclx_odr_to_freq_val(odr);
+
+	return 0;
+}
+
+#ifdef IIS2ICLX_ACCEL_ODR_RUNTIME
+static int iis2iclx_accel_odr_set(const struct device *dev, uint16_t freq)
+{
+	int odr;
+
+	odr = iis2iclx_freq_to_odr_val(freq);
+	if (odr < 0) {
+		return odr;
+	}
+
+	if (iis2iclx_accel_set_odr_raw(dev, odr) < 0) {
+		LOG_DBG("failed to set accelerometer sampling rate");
+		return -EIO;
+	}
+
+	return 0;
+}
+#endif
+
+#ifdef IIS2ICLX_ACCEL_FS_RUNTIME
+static int iis2iclx_accel_range_set(const struct device *dev, int32_t range)
+{
+	int fs;
+	struct iis2iclx_data *data = dev->data;
+
+	fs = iis2iclx_accel_range_to_fs_val(range);
+	if (fs < 0) {
+		return fs;
+	}
+
+	if (iis2iclx_accel_set_fs_raw(dev, fs) < 0) {
+		LOG_DBG("failed to set accelerometer full-scale");
+		return -EIO;
+	}
+
+	data->acc_gain = (iis2iclx_accel_fs_sens[fs] * GAIN_UNIT_XL);
+	return 0;
+}
+#endif
+
+static int iis2iclx_accel_config(const struct device *dev,
+				   enum sensor_channel chan,
+				   enum sensor_attribute attr,
+				   const struct sensor_value *val)
+{
+	switch (attr) {
+#ifdef IIS2ICLX_ACCEL_FS_RUNTIME
+	case SENSOR_ATTR_FULL_SCALE:
+		return iis2iclx_accel_range_set(dev, sensor_ms2_to_g(val));
+#endif
+#ifdef IIS2ICLX_ACCEL_ODR_RUNTIME
+	case SENSOR_ATTR_SAMPLING_FREQUENCY:
+		return iis2iclx_accel_odr_set(dev, val->val1);
+#endif
+	default:
+		LOG_DBG("Accel attribute not supported.");
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int iis2iclx_attr_set(const struct device *dev,
+			       enum sensor_channel chan,
+			       enum sensor_attribute attr,
+			       const struct sensor_value *val)
+{
+	switch (chan) {
+	case SENSOR_CHAN_ACCEL_XYZ:
+		return iis2iclx_accel_config(dev, chan, attr, val);
+#if defined(CONFIG_IIS2ICLX_SENSORHUB)
+	case SENSOR_CHAN_MAGN_XYZ:
+	case SENSOR_CHAN_PRESS:
+	case SENSOR_CHAN_HUMIDITY:
+		return iis2iclx_shub_config(dev, chan, attr, val);
+#endif /* CONFIG_IIS2ICLX_SENSORHUB */
+	default:
+		LOG_WRN("attr_set() not supported on this channel.");
```

Comment:
There is some inconsistency using LOG_WRN vs. LOG_ERR when returning an errno throughout the driver. Is this intended?

Reply:
Mmmmh, no I think that are leftover from copying paisting from previous driver.
Idea should be (I guess):

   - use LOG_ERR when returning error
   - use LOG_WRN/LOG_INFO otherwise